lib/spack/spack/binary_distribution.py

```diff
@@ -198,8 +198,14 @@ def _associate_built_specs_with_mirror(self, cache_key, mirror_url):
             with self._index_file_cache.read_transaction(cache_key):
                 db._read_from_file(cache_path)
 
+            # All specs for which tarballs are available
             spec_list = db.query_local(installed=False, in_buildcache=True)
 
+            # and externals.
+            spec_list += [
+                s for s in db.query_local(installed=False, in_buildcache=False) if s.external
```

Comment:
I don't think that `installed=False` is what we want here, I think we want `installed=any`. The external can be reused whether or not it was "installed".

It would also be good if we could restrict this to externals that are depended on by one of the packages in `spec_list`.

Reply:
`query_local` isn't super helpful :) I've turned it into a single call to that with extra filtering afterwards.

> It would also be good if we could restrict this to externals that are depended on by one of the packages in `spec_list`.

I think this is a bit too niche to deal with

Reply:
This should only consider externals in `possible_dependencies` for the current solve.